### ISSUE TITLE
Swiss Keyboard not automatically selected by dosbox-staging, but it works in original Dosbox

### ISSUE THREAD

**Issue by cwacha** [dosbox-staging/dosbox-staging#3523]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### What Operating System are you using?

Windows 11

### Relevant hardware info

not relevant

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

My Windows is configured to use english language for display and swiss german keyboard as the input type. 

The original Dosbox 0.74-3 gets this right. It automatically load the SG keyboard driver. Howerver, dosbox-staging loads the german driver.  I have to manually change it using "keyb sg" (which I can do but I wonder why the automation does not work.)

```
2024-03-21 09:35:39.487 | DOS: Loaded codepage 858 for detected language 'de'
2024-03-21 09:35:39.519 | DOS: Loaded modern locale for country 49, 'Germany'
```

I guess you want more information, please let me know what exactly you need.

### Steps to reproduce the behaviour.

Explain how to reproduce
1. Set language to english, set keyboard to CH
2. launch dosbox-staging 0.81.0
3. notice that language is now german.

### Your configuration

```ini
everything on defaults.
```

### Provide a Log

```text
2024-03-21 09:46:50.463 | arguments: C:\\Users\\xxx\\Downloads\\dosbox-staging-windows-v0.81.0\\dosbox-staging-v0.81.0\\dosbox.exe
2024-03-21 09:46:50.464 | Current dir: C:\Users\xxx\Desktop
2024-03-21 09:46:50.464 | stderr verbosity: 0
2024-03-21 09:46:50.464 | -----------------------------------
2024-03-21 09:46:50.464 | dosbox-staging version 0.81.0
2024-03-21 09:46:50.464 | ---
2024-03-21 09:46:50.465 | LOG: Loguru version 2.1.0 initialised
2024-03-21 09:46:50.465 | CONFIG: Using portable configuration layout in 'C:\Users\xxx\Downloads\dosbox-staging-windows-v0.81.0\dosbox-staging-v0.81.0'
2024-03-21 09:46:50.471 | CONFIG: Loaded primary config file 'C:\Users\xxx\Downloads\dosbox-staging-windows-v0.81.0\dosbox-staging-v0.81.0\dosbox-staging.conf'
2024-03-21 09:46:50.477 | LANG: Loaded language file: C:\Users\xxx\Downloads\dosbox-staging-windows-v0.81.0\dosbox-staging-v0.81.0\resources\translations\de.lng
2024-03-21 09:46:50.515 | SDL: version 2.26.5 initialised (windows video and wasapi audio)
2024-03-21 09:46:50.515 | DISPLAY: Initialised 1422x1066 windowed mode using bilinear scaling on display-0
2024-03-21 09:46:50.708 | OPENGL: Not requesting an sRGB framebuffer because Intel's driver is unreliable
2024-03-21 09:46:50.897 | OPENGL: Vendor: Intel
2024-03-21 09:46:50.952 | OPENGL: Version: 4.6.0 - Build 27.20.100.8854
2024-03-21 09:46:50.952 | OPENGL: GLSL version: 4.60 - Build 27.20.100.8854
2024-03-21 09:46:50.952 | OPENGL: NPOT textures supported
2024-03-21 09:46:50.957 | RENDER: Using shader 'none'
2024-03-21 09:46:50.992 | VGA: Pixel doubling enabled
2024-03-21 09:46:50.992 | VIDEO: Using the DOS video mode's frame rate
2024-03-21 09:46:51.001 | MEMORY: Using 4096 DOS memory pages (16 MB) at address: 0000016B8DA18040
2024-03-21 09:46:51.026 | VGA: Double scanning VGA video modes enabled
2024-03-21 09:46:51.035 | VIDEO: Initialised S3 Trio64 VESA 2.0 compatible with 4 MB of FP DRAM supporting 86 modes
2024-03-21 09:46:51.041 | VOODOO: Initialized with 4 MB of RAM, multithreading, and no bilinear filtering
2024-03-21 09:46:51.053 | CAPTURE: Image capturer started
2024-03-21 09:46:51.301 | MIXER: Negotiated 2-channel 48000 Hz audio of 1024-frame blocks
2024-03-21 09:46:51.318 | MIXER: Master compressor enabled
fluidsynth: warning: Some invalid sample loops were sanitized! If you experience audible glitches, start fluidsynth in verbose mode for detailed information.
2024-03-21 09:46:51.566 | FSYNTH: Using SoundFont 'C:\Users\xxx\Downloads\dosbox-staging-windows-v0.81.0\dosbox-staging-v0.81.0\soundfonts\SC-55.sf2'
2024-03-21 09:46:51.568 | FSYNTH: Chorus enabled with 3 voices at level 1.20, 0.30 Hz speed, 8.00 depth, and sine-wave modulation
2024-03-21 09:46:51.573 | FSYNTH: Reverb enabled with a 0.61 room size, 0.23 damping, 0.76 width, and level 0.56
2024-03-21 09:46:51.574 | FSYNTH: Operating at 48000 Hz without resampling
2024-03-21 09:46:51.574 | MIDI: Opened device: fluidsynth
2024-03-21 09:46:51.574 | MPU-401: Running in intelligent mode on ports 330h and 331h
2024-03-21 09:46:51.574 | OPL: Operating at 48000 Hz without resampling
2024-03-21 09:46:51.574 | OPL: Running OPL3 on ports 220h and 388h
2024-03-21 09:46:51.574 | SB16: Sound Blaster 16 OPL output filter enabled
2024-03-21 09:46:51.575 | DMA: Initialised primary controller
2024-03-21 09:46:51.575 | DMA: Initialised secondary controller
2024-03-21 09:46:51.575 | SB: Operating at 22050 Hz and upsampling to the output rate
2024-03-21 09:46:51.575 | SB16: Modern DAC output filter enabled
2024-03-21 09:46:51.578 | SB16: Setting 'BLASTER' environment variable to 'A220 I7 D1 H5 T6'
2024-03-21 09:46:51.578 | SB16: Running on port 220h, IRQ 7, DMA 1, and high DMA 5
2024-03-21 09:46:51.578 | PCSPEAKER: Operating at 48000 Hz without resampling
2024-03-21 09:46:51.578 | PCSPEAKER: Initialised discrete model
2024-03-21 09:46:51.578 | PCSPEAKER: Highpass filter enabled (18 dB/oct at 120 Hz)
2024-03-21 09:46:51.579 | PCSPEAKER: Lowpass filter enabled (12 dB/oct at 4800 Hz)
2024-03-21 09:46:51.580 | MOUSE: Will be captured after the first left or right button click
2024-03-21 09:46:51.582 | MOUSE: Middle button will capture/release the mouse (clicks not sent to the game/program)
2024-03-21 09:46:51.583 | MOUSE (PS/2): Standard, 3 buttons
2024-03-21 09:46:51.602 | SDL: Using standard SDI (auto) display refresh rate of 59.94 Hz
2024-03-21 09:46:51.652 | DISPLAY: VGA 720x400 16-colour text mode 03h at 70.087 Hz throttled VFR, scaled to 1421x1066 pixels with 1:1.35 (20:27) pixel aspect ratio
2024-03-21 09:46:51.739 | DOS: Loaded codepage 858 for detected language 'de'
2024-03-21 09:46:51.750 | DOS: Loaded modern locale for country 49, 'Germany'
2024-03-21 09:46:51.750 | SLIRP: Slirp version: 4.7.0
2024-03-21 09:46:51.752 | SLIRP: Successfully initialized
2024-03-21 09:46:51.752 | NE2000: Initialised on port 300h and IRQ 3
2024-03-21 09:46:51.757 | CONFIG: Loaded auto-mounted drive config file 'C:\Users\xxx\Downloads\dosbox-staging-windows-v0.81.0\dosbox-staging-v0.81.0\resources\drives\y.conf'
2024-03-21 09:46:52.008 | MAPPER: No joysticks found
2024-03-21 09:46:52.009 | MAPPER: Loaded default key bindings
2024-03-21 09:46:52.014 | MOUNT: Path '..\..\..\Downloads\dosbox-staging-windows-v0.81.0\dosbox-staging-v0.81.0\resources\drives\y' found
2024-03-21 09:46:52.016 | UNICODE: No fallback mapping for code point 0x2665
```

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

**Comment by cwacha** [comment]
You can see that the first preloaded language is US (409)
![image](https://github.com/dosbox-staging/dosbox-staging/assets/23018497/fb61b55e-0f9b-4f9e-bdab-20fff4790c72)

But Substitutes map this correctly to SG
![image](https://github.com/dosbox-staging/dosbox-staging/assets/23018497/92bcf91b-9e7b-4182-b6c8-d96d38cb96ce)

![image](https://github.com/dosbox-staging/dosbox-staging/assets/23018497/88159bf1-c97b-4856-acb2-b82c9fe74d97)

**Comment by FeralChild64** [comment]
@cwacha Could you paste also the screenshot of your language/keyboard settings, from the Settings application?

**Comment by cwacha** [comment]
Will do later. Could you send me a link to a dosbox with debug enabled? Then I can send you the debug output. I saw in the code that you are doing this via the language code which for ch would be de_ch. And if you drop the second part you will end up with de. Which is a valid language in ch but not the right keyboard layout. The sg layout has some French letters in there which the de layout does have. But the Swiss French I believe just use the French layout. So there it works. The annoying part is that in the sg layout the * and a couple others are not in the same place as I the de layout. So to get this right in windows you must follow the registry or use the corresponding .net call.

**Comment by cwacha** [comment]
@FeralChild64 I am using an english windows with english language configured, but I also want swiss german keyboard and especially the swiss language settings (date format etc.). The only way to select the German Switzerland Regional Format is by installing German Switzerland as language. Hence you see two languages active (In Windows you cannot have one language with the language settings from another language, other than in Linux or in Mac OS). The annoying thing is then that you always get the keyboard picker in the system tray. Thats why both languages have the SG keyboard configured. Why dosbox is seeing german language I don't know. It should be english.

![image](https://github.com/dosbox-staging/dosbox-staging/assets/23018497/fb99a2e8-8354-4b8c-84f9-8405f0a691dc)
![image](https://github.com/dosbox-staging/dosbox-staging/assets/23018497/1fd1655e-b48e-4392-99e6-e3a4e6d9861b)

**Comment by FeralChild64** [comment]
Great, the screenshot will be very helpful. I don't need a debug logs for now, thank you.

In short: the whole language detection code is oversimplified and has many problems (see https://github.com/dosbox-staging/dosbox-staging/issues/3498, https://github.com/dosbox-staging/dosbox-staging/issues/3490, or https://github.com/dosbox-staging/dosbox-staging/issues/1925) - I intend to rewrite it. Basically, we need to establish the following locale settings, either using auto-detection or from `dosbox-staging.conf` file:

1. Messages (language)
2. DOS code page
3. DOS keyboard layout
4. DOS country ID
5. Numeric format
6. Time/Date format
7. Currency
8. Collation (sorting) rules - we do not currently provide locale-dependent collation tables to DOS programs, but I intend to handle this someday

DOS code page is going to be determined from the keyboard layout and locale era (modern locale = prefer codepage with € sign for certain layouts).

For POSIX systems, we have environment variables: `LANG`, `LANGUAGE`, `LC_ALL` (these are the generic ones), `LC_MESSAGES`, `LC_NUMERIC`, `LC_TIME`, `LC_MONETARY`, `LC_COLLATE`, etc. - they have to be evaluated with specific order. Keyboard layout seems to be harder, I'm not sure if there is a way to detect it if using Wayland.

Numeric format, time/date format and currency can be taken from detected country ID, if the host OS does not provide enough granularity.

For Windows, I hope `GetUserDefaultUILanguage` / `GetKeyboardLayout` / `GetUserDefaultLocaleName` will allow me to detect enough.

For MacOS X, there also seems to be some API available - `CFLocaleCopyCurrent` / `CFLocaleGetValue` / `kCFLocaleCountryCode` / `kCFLocaleLanguageCode` / etc. I don't have a MacOS X, unfortunately, but the API looks simple.

----

As you see, this whole mess won't be cleaned up within a day or two :)

**Comment by johnnovak** [comment]
> For MacOS X, there also seems to be some API available - `CFLocaleCopyCurrent` / `CFLocaleGetValue` / `kCFLocaleCountryCode` / `kCFLocaleLanguageCode` / etc. I don't have a MacOS X, unfortunately, but the API looks simple.

Feel free to ask me to help with testing this on macOS. I'm sure @kklobe, @Burrito78 (German) and @Grounded0 (Finnish) would be happy to lend a hand as well.

**Comment by cwacha** [comment]
@FeralChild64 I can also test on MacOS. Testing different keyboard layouts there is almost idiot safe and does not even need a reboot.

Thanks for the very quick turnaround. This does not need to be fixed in one day as setting they keyboard manually in dosbox-staging.conf is simple and is a perfect workaround.

**Comment by johnnovak** [comment]
> @FeralChild64 I can also test on MacOS. Testing different keyboard layouts there is almost idiot safe and does not even need a reboot.

Thanks for offering your help on @cwacha, much appreciated.

**Comment by NicknineTheEagle** [comment]
@FeralChild64 I'd like to point out that using host keyboard layout for anything is not a good idea since there are countries like Russia where dual keyboard layout is the standard. Furthermore, user may add additional layouts to the system for any number of reasons which can further mess with potential detection process.

**Comment by rderooy** [comment]
I would think using host keyboard layout by default (as detected on DOSBox startup), should be correct in almost all use cases. And should be better than any other "default". Although yes, you can dynamically switch host layout, and DOSBox would be none the wiser.

e..g.,

Where I live there are 4 keyboards in common use
- Swiss French (commonly used for public administration, public schools and uni)
- French (readily available in shops)
- Belgium (readily available in shops)
- German (slightly less common)

And I don't use any of the above. Instead I use a US layout.

**Comment by FeralChild64** [comment]
Well, the keyboard layout auto-detection is never going to work 100% reliably. I have already spent quite some time comparing FreeDOS layouts to Linux/X11 ones, I'm not mapping the layouts blindly - I don't want to auto-set a QWERTZ or AZERTY layout inside the emulation if the host uses QWERTY, and I don't want to map `gr` layout to `gr` layout, as _gr_ stands for _GReek_ in Linux (see the norm ISO 3166-1 alpha-2), but _GeRman_ in FreeDOS - hopefully I'll ba able to come up with something good enough. For unknown layouts, US English will be used.

In case of multiple keyboard layout - on Linux the main (primary) one is going to be used, I don't even know how to check which layout is currently in use. I didn't investigate Windows or macOS yet.

**Comment by NicknineTheEagle** [comment]
> I would think using host keyboard layout by default (as detected on DOSBox startup), should be correct in almost all use cases. And should be better than any other "default". Although yes, you can dynamically switch host layout, and DOSBox would be none the wiser.
> 
> e..g.,
> 
> Where I live there are 4 keyboards in common use
> 
>     * Swiss French (commonly used for public administration, public schools and uni)
> 
>     * French (readily available in shops)
> 
>     * Belgium (readily available in shops)
> 
>     * German (slightly less common)
> 
> 
> And I don't use any of the above. Instead I use a US layout.

Russia uses US QWERTY and Russian ЙЦУКЕН in tandem and either one of them can be active at any given time since do need both of them. US layout is used to type Latin characters and Russian layout is used to type Cyrillic characters and people need to switch between them constantly with Alt-Shift.
Why use host layout at all? Why not use locale instead?

**Comment by NicknineTheEagle** [comment]
> In case of multiple keyboard layouts - on Linux the main (primary) one is going to be used

That is also far from reliable. In Russia, it's not at all guaranteed for Russian layout to be the default one - depending on user's preferences, they may have US layout as the default (I'd imagine most programmers do, for example).

**Comment by FeralChild64** [comment]
> Why use host layout at all? Why not use host locale instead?

Because host locale setting is not precise enough.
Let’s assume the locale is US for everything. You can now have a standard US keyboard layout, US International layout, a Dvorak layout, or a Colemak layout.
For Polish you can have a so-called Programmers layout, or a Typewritter layout (very rare nowadays, but still supported by both Linux and FreeDOS keyboard layout set).
And so on…

**Comment by rderooy** [comment]
> Why use host layout at all? Why not use host locale instead?

Because that would obviously be wrong. The host keyboard is what should be used. The locale does not guarantee the keyboard layout in any conceivable way.

I just explained that where I live there are at least 4 keyboard layouts in common use. Another example is the Netherlands, where the official layout would be a Dutch layout right? Well, yes, unless you buy a computer (or keyboard) in ANY high-street computer store and find that it has a US layout instead. To get a Dutch layout you pretty much need to special order one (they are used in public offices).

**Comment by FeralChild64** [comment]
Probably not really feasible. On Linux, I don't even have any easy way to check which of the keyboard layouts is currently selected, not to mention receiving information when it gets changed.

**Comment by johnnovak** [comment]
Umm, use Staging like you would real MS-DOS, @Torinde? Pick one layout and live with it? It's the authentic experience 😆

**Comment by FeralChild64** [comment]
Even if we have a way to detect locale change when running, there remains w problem with the guest software: how will, for example, Windows 3.x react on the keyboard layout (and possibly code page) change magically?

BTW, Yesterday evening I have learned that many FreeDOS keyboard layouts are dual-layer, or even triple-layer, see `https://gitlab.com/FreeDOS/base/keyb_lay/-/blob/master/DOC/KEYB/LAYOUTS/LAYOUTS.TXT?ref_type=heads`. At least for some of these layouts, the `[ALT]+[LEFT_SHIFT]` / `[ALT]+[RIGHT_SHIFT]` / `[ALT]+[LEFT_CTRL]` shortcuts work within the DOSBox Staging.

Maybe someone someday will write the layout-switch synchronization, but one thing is sure: I don't intend to do this in a foreseeable future. My first locale-related commit is going to be rather large (but mainly due to a lot of static data) even without such deep host OS integration. Thankfully, I have found some web pages with keyboard layout visualizations for Windows and macOS (on Linux I can just click "Preview" in the keyboard layout tool):

https://kbdlayout.info
https://keyshorts.com/blogs/blog/37615873-how-to-identify-macbook-keyboard-localization

**Comment by johnnovak** [comment]
I'm glad Hungarian uses the Latin alphabet and 99.99% of the the time you can understand the text even if all accented letters are turned to non-accented... The Russian situation sounds like a nightmare @Torinde.

Like many nerd types from the 80s, I only use the Hungarian layout if I must write something formal or something to be printed, etc.

**Comment by FeralChild64** [comment]
... and since I nevertheless had to add some huge tables with keyboard layouts metadata...

![Screenshot2](https://github.com/dosbox-staging/dosbox-staging/assets/48332137/cbc0caf2-5bff-4ada-9138-73befbccdc51)

![Screenshot1](https://github.com/dosbox-staging/dosbox-staging/assets/48332137/0ece8626-a21d-4b5f-afc2-7ffa20d12a88)

**Comment by NicknineTheEagle** [comment]
> I'm glad Hungarian uses the Latin alphabet and 99.99% of the the time you can understand the text even if all accented letters are turned to non-accented... The Russian situation sounds like a nightmare

Eh, you get used to it. The far bigger problem is that some programs don't handle Cyrillic paths properly. Have fun if you happened to name your Windows account with Cyrillic characters on WinXP.

**Comment by NicknineTheEagle** [comment]
> > sounds like a nightmare
> 
> To visualize it take a look at #3239 :) That's why I suggest KEYB to support `LeftShift` + `RightShift` as combo for on-the-fly codepage switching between 437 and whatever the _other_ loaded codepage is.
> 
> > The KEYB command output
> 
> I like it - providing that expanded info would be good!

That sounds like a tangentially related issue, we're talking keyboard layouts here.

**Comment by cwacha** [comment]
Not sure I know what I do. I downloaded this artifact from the job [dosbox-staging-windows-x64-0.82.0-alpha-dc16a](https://github.com/dosbox-staging/dosbox-staging/actions/runs/10548517120/artifacts/1852008973). Then I extracted it and ran it without making any modifications to the config file.

Result looks great. I would say it worked. Let me try this on my Mac

```
2024-09-30 22:19:38.949 | DOS: Country detected from 'de-CH'
2024-09-30 22:19:38.949 | DOS: Using currency format detected from host OS
2024-09-30 22:19:38.949 | DOS: Loaded locale for country 61 'Australia'
2024-09-30 22:19:38.949 | DOS: Keyboard layout detected from '00000409->00000807;00000409;00000807'
2024-09-30 22:19:38.957 | DOS: Loaded code page 850
2024-09-30 22:19:38.959 | DOS: Loaded keyboard layout 'sd' - 'Swiss (German, QWERTZ)'
2024-09-30 22:19:38.960 | SLIRP: Slirp version: 4.8.0
```

![image](https://github.com/user-attachments/assets/599be1ec-f9d3-4642-97f4-91a4042cc42d)

**Comment by FeralChild64** [comment]
I'm glad it worked for you :)

It won't be a part of the 0.82.0, unfortunately - it is a huge change, and we are at the end of 0.82.0 development, in a release candidate phase.

**Comment by johnnovak** [comment]
> I'm glad it worked for you :)
> 
> It won't be a part of the 0.82.0, unfortunately - it is a huge change, and we are at the end of 0.82.0 development, in a release candidate phase.

Yeah, definitely 0.83.x material. Thanks for the testing @cwacha , much appreciated.

**Comment by ulangmeier** [comment]
#TLDR
Quick fix for Swiss German Keyboard layout:

Go to the DosBox Options category [dos] and set 

keyboardlayout from auto to sg (swiss german)
